packages/plugins/other/visitor-plugin-common/src/selection-set-to-object.ts

```diff
@@ -314,29 +339,64 @@ export class SelectionSetToObject<Config extends ParsedDocumentsConfig = ParsedD
     const possibleTypes = getPossibleTypes(this._schema, this._parentSchemaType);
 
     if (!this._config.mergeFragmentTypes || this._config.inlineFragmentTypes === 'mask') {
-      const grouped = possibleTypes.reduce((prev, type) => {
+      const grouped = possibleTypes.reduce<GroupedStringifedTypes>((prev, type) => {
         const typeName = type.name;
         const schemaType = this._schema.getType(typeName);
 
         if (!isObjectType(schemaType)) {
           throw new TypeError(`Invalid state! Schema type ${typeName} is not a valid GraphQL object!`);
         }
 
-        const selectionNodes = selectionNodesByTypeName.get(typeName) || [];
+        const allNodes = selectionNodesByTypeName.get(typeName) || [];
 
         prev[typeName] ||= [];
 
+        const { incrementalNodes, selectionNodes } = allNodes.reduce<{
+          selectionNodes: (SelectionNode | FragmentSpreadUsage)[];
+          incrementalNodes: FragmentSpreadUsage[];
+        }>(
+          (acc, node) => {
+            if ('fragmentDirectives' in node && hasIncrementalDeliveryDirectives(node.fragmentDirectives)) {
+              acc.incrementalNodes.push(node);
+            } else {
+              acc.selectionNodes.push(node);
+            }
+            return acc;
+          },
+          { selectionNodes: [], incrementalNodes: [] }
+        );
+
         const { fields } = this.buildSelectionSet(schemaType, selectionNodes);
-        const transformedSet = this.selectionSetStringFromFields(fields);
 
+        const transformedSet = this.selectionSetStringFromFields(fields);
         if (transformedSet) {
           prev[typeName].push(transformedSet);
         } else {
           mustAddEmptyObject = true;
         }
 
+        for (const incrementalNode of incrementalNodes) {
+          if (this._config.inlineFragmentTypes === 'mask' && 'fragmentName' in incrementalNode) {
+            const { fields: incrementalFields } = this.buildSelectionSet(schemaType, [incrementalNode], {
+              unsetTypes: true,
+            });
+            prev[typeName].push(this.selectionSetStringFromFields(incrementalFields));
+
+            continue;
+          }
+          const { fields: initialFields } = this.buildSelectionSet(schemaType, [incrementalNode]);
+          const { fields: subsequentFields } = this.buildSelectionSet(schemaType, [incrementalNode], {
+            unsetTypes: true,
+          });
+
+          const initialSet = this.selectionSetStringFromFields(initialFields);
+          const subsequentSet = this.selectionSetStringFromFields(subsequentFields);
+
+          prev[typeName].push({ union: [initialSet, subsequentSet] });
+        }
```

Comment:
For deferred fragments, I'm creating a union of initial fields and deferred fields, for example: `{name: string} | {name?: never}`.

In the case of `inlineFragmentTypes: mask`, I don't want a union because that would generate `{ ' $fragmentRefs'?: { FilmItemFragment: FilmItemFragment } } l { ' $fragmentRefs'?: { FilmItemFragment: FilmItemFragment } };` (union with the same fields). In such case, `buildSelectionSet` adds an `Incremental` generic type: https://github.com/dotansimha/graphql-code-generator/pull/9196/files#diff-062404aa39a1345b9a8d7b97f61efd8d516ceb3ccaf8e637a0e1e18b9c9f4fcaR657